esphome/components/remote_base/nec_protocol.cpp

```
@@ -6,93 +6,138 @@ namespace remote_base {
+      .command_repeats = 0, // Start with 0, as the first frame is counted explicitly
   };
-  if (!src.expect_item(HEADER_HIGH_US, HEADER_LOW_US))
+
+  // Validate the AGC header (start of frame)
+  if (!src.expect_item(HEADER_HIGH_US, HEADER_LOW_US)) {
     return {};
+  }
 
+  // Decode 16-bit Address
   for (uint16_t mask = 1; mask; mask <<= 1) {
     if (src.expect_item(BIT_HIGH_US, BIT_ONE_LOW_US)) {
-      data.address |= mask;
+      data.address |= mask;  // Logic '1'
     } else if (src.expect_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-      data.address &= ~mask;
+      data.address &= ~mask; // Logic '0'
     } else {
       return {};
     }
   }
 
+  // Decode 16-bit Command
   for (uint16_t mask = 1; mask; mask <<= 1) {
     if (src.expect_item(BIT_HIGH_US, BIT_ONE_LOW_US)) {
-      data.command |= mask;
+      data.command |= mask;  // Logic '1'
     } else if (src.expect_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-      data.command &= ~mask;
+      data.command &= ~mask; // Logic '0'
     } else {
       return {};
     }
   }
 
-  while (src.peek_item(BIT_HIGH_US, BIT_ONE_LOW_US) || src.peek_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-    uint16_t command = 0;
-    for (uint16_t mask = 1; mask; mask <<= 1) {
-      if (src.expect_item(BIT_HIGH_US, BIT_ONE_LOW_US)) {
-        command |= mask;
-      } else if (src.expect_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-        command &= ~mask;
-      } else {
-        return {};
-      }
+  // Validate the final mark (end of the message frame)
+  if (!src.expect_mark(BIT_HIGH_US)) {
+    return {};
+  }
+
+  // Check for inter-frame space (only if repeats are expected)
+  if (src.peek_space(SPACE_INTER_FRAME_US)) {
+    src.expect_space(SPACE_INTER_FRAME_US); // Consume inter-frame space
+  }
+
+  // Decode AGC Repeat Codes
+  while (src.peek_item(HEADER_HIGH_US, HEADER_LOW_US / 2)) {
+    // Validate the repeat header
+    if (!src.expect_item(HEADER_HIGH_US, HEADER_LOW_US / 2)) {
+      break;
```

Breaking the loop instead of returning invalid NEC data here makes sense because a failure to validate a repeat header doesn’t necessarily invalidate the entire transmission. The main frame has already been successfully decoded, so the NEC data is still valid even if some repeat codes are malformed or missing. This approach ensures partial but meaningful data can still be returned.